File: xabber/src/main/java/com/xabber/android/data/extension/reliablemessagedelivery/TimeElement.java
package com.xabber.android.data.extension.reliablemessagedelivery;

import org.jivesoftware.smack.packet.ExtensionElement;
import org.jivesoftware.smack.util.XmlStringBuilder;

public class TimeElement implements ExtensionElement {

    public static final String NAMESPACE = null;
    public static final String ELEMENT = "time";
    public static final String ATTRIBUTE_BY = "by";
    public static final String ATTRIBUTE_STAMP = "stamp";

    private String by = null;
    private String stamp = null;

    public TimeElement(String by, String timeStamp) {
        this.by = by;
        this.stamp = timeStamp;
    }

    public TimeElement() { }

    @Override
    public String getNamespace() {
        return NAMESPACE;
    }

    @Override
    public String getElementName() {
        return ELEMENT;
    }

    public String getBy() {
        return by;
    }

    public void setBy(String by) {
        this.by = by;
    }

    public String getStamp() {
        return stamp;
    }

    public void setStamp(String stamp) {
        this.stamp = stamp;
    }

    @Override
    public XmlStringBuilder toXML() {
        XmlStringBuilder xmlStringBuilder = new XmlStringBuilder(this);
        xmlStringBuilder.attribute(ATTRIBUTE_BY, by);
        xmlStringBuilder.attribute(ATTRIBUTE_STAMP, stamp);
        xmlStringBuilder.closeEmptyElement();
        return xmlStringBuilder;
    }

}

File: xabber/src/main/java/com/xabber/android/data/extension/reliablemessagedelivery/TimeElement.kt
package com.xabber.android.data.extension.reliablemessagedelivery

import com.xabber.android.data.extension.reliablemessagedelivery.TimeElement.Companion.ELEMENT
import com.xabber.android.data.extension.reliablemessagedelivery.TimeElement.Companion.NAMESPACE
import org.jivesoftware.smack.packet.ExtensionElement
import org.jivesoftware.smack.packet.Stanza
import org.jivesoftware.smack.util.XmlStringBuilder

class TimeElement(val by: String,
                  val timeStamp: String,
): ExtensionElement {

    override fun getNamespace() = NAMESPACE

    override fun getElementName() = ELEMENT

    override fun toXML() = XmlStringBuilder(this).apply {
        attribute(ATTRIBUTE_BY, by)
        attribute(ATTRIBUTE_STAMP, timeStamp)
        closeEmptyElement()
    }

    companion object {
        const val NAMESPACE = DeliveryManager.NAMESPACE
        const val ELEMENT = "time"
        const val ATTRIBUTE_BY = "by"
        const val ATTRIBUTE_STAMP = "stamp"
    }

}

fun Stanza.hasTimeElement() = this.hasExtension(ELEMENT, NAMESPACE)
fun Stanza.getTimeElement() = this.getExtension<TimeElement>(ELEMENT, NAMESPACE)